File: js/src/typesafe/answers.ts
import { TypeSafeError } from '@typesafe-ai/sdk';
import type { Content } from './questions.js';

/**
 * One question's answer. The SDK hands back plain objects; these wrap them so a
 * caller can ask what kind of answer it is, and so a shape mismatch surfaces
 * here rather than as a missing field further down
 */
export abstract class Answer {
    constructor(protected readonly data: Record<string, unknown>) {}

    /** The discriminator the API sent */
    abstract type(): string;

    toObject(): Record<string, unknown> {
        return this.data;
    }

    /** The answer class matching the payload's `type`, or `null` for one not modelled here */
    static fromObject(data: Record<string, unknown>): Answer | null {
        switch (data['type']) {
            case 'noul':
                return new NoulAnswer(data);
            case 'choice':
                return new ChoiceAnswer(data);
            case 'score':
                return new ScoreAnswer(data);
            default:
                return null;
        }
    }

    protected number(key: string): number {
        const value = this.data[key];

        if (typeof value !== 'number') {
            throw new TypeSafeError(`Expected a number for "${key}" in a ${this.type()} answer.`);
        }

        return value;
    }

    protected numberMap(key: string): Map<string, number> {
        const values = this.data[key] ?? {};

        if (typeof values !== 'object' || values === null) {
            throw new TypeSafeError(`Expected a map for "${key}" in a ${this.type()} answer.`);
        }

        const map = new Map<string, number>();

        for (const [name, value] of Object.entries(values as Record<string, unknown>)) {
            map.set(name, typeof value === 'number' ? value : Number(value) || 0);
        }

        return map;
    }
}

/** A yes/no answer */
export class NoulAnswer extends Answer {
    type(): string {
        return 'noul';
    }

    /** Probability of a yes answer, from zero to one */
    noul(): number {
        return this.number('noul');
    }
}

/** The label the model selected, with its probability across every label */
export class ChoiceAnswer extends Answer {
    type(): string {
        return 'choice';
    }

    choice(): string {
        const choice = this.data['choice'];

        return typeof choice === 'string' ? choice : '';
    }

    confidence(): number {
        return this.number('confidence');
    }

    probabilities(): Map<string, number> {
        return this.numberMap('probabilities');
    }

    /** The probability of one label, or `null` when the model did not report it */
    probabilityOf(label: string): number | null {
        return this.probabilities().get(label) ?? null;
    }
}

/** An expected score with the rubric it was drawn from */
export class ScoreAnswer extends Answer {
    type(): string {
        return 'score';
    }

    /** Expected score, which may fall between the integer rubric levels */
    score(): number {
        return this.number('score');
    }

    nearestLevel(): number {
        return Math.round(this.score());
    }

    confidence(): number {
        return this.number('confidence');
    }

    legend(): Map<number, Content> {
        const legend = this.data['legend'] ?? {};

        if (typeof legend !== 'object' || legend === null) {
            return new Map();
        }

        const keyed = new Map<number, Content>();

        for (const [score, description] of Object.entries(legend as Record<string, Content>)) {
            keyed.set(Number(score), description);
        }

        return keyed;
    }
}

/**
 * Token counts for a request, when the API reports them.
 *
 * A provider that reports none is not a call that cost nothing, so the two are
 * kept apart and the report says how many calls it could not price
 */
export class Usage {
    private constructor(
        public readonly inputTokens: number | null,
        public readonly outputTokens: number | null,
    ) {}

    static fromObject(data: Record<string, unknown> | undefined): Usage {
        return new Usage(asNumber(data?.['input_tokens']), asNumber(data?.['output_tokens']));
    }

    /** Input and output tokens combined, or `null` when neither was reported */
    totalTokens(): number | null {
        if (this.inputTokens === null && this.outputTokens === null) {
            return null;
        }

        return (this.inputTokens ?? 0) + (this.outputTokens ?? 0);
    }
}

function asNumber(value: unknown): number | null {
    return typeof value === 'number' && Number.isFinite(value) ? value : null;
}

/** Answers keyed by question name, with the model that produced them and the tokens they cost */
export class SystemOneResponse {
    private constructor(
        public readonly model: string,
        private readonly byName: Map<string, Answer>,
        public readonly usage: Usage,
    ) {}

    static fromObject(data: Record<string, unknown>): SystemOneResponse {
        const byName = new Map<string, Answer>();
        const raw = data['answers'];

        if (typeof raw === 'object' && raw !== null) {
            for (const [name, answer] of Object.entries(raw as Record<string, unknown>)) {
                if (typeof answer !== 'object' || answer === null) {
                    continue;
                }

                // An answer type a later API adds is left out rather than fatal.
                const built = Answer.fromObject(answer as Record<string, unknown>);

                if (built !== null) {
                    byName.set(name, built);
                }
            }
        }

        const model = data['model'];
        const usage = data['usage'];

        return new SystemOneResponse(
            typeof model === 'string' ? model : '',
            byName,
            Usage.fromObject(typeof usage === 'object' && usage !== null ? usage as Record<string, unknown> : undefined),
        );
    }

    answers(): Map<string, Answer> {
        return this.byName;
    }

    answer(name: string): Answer {
        const answer = this.byName.get(name);

        if (answer === undefined) {
            throw new TypeSafeError(
                `No answer named "${name}" in the response; got: ${this.byName.size === 0 ? 'none' : [...this.byName.keys()].join(', ')}.`,
            );
        }

        return answer;
    }

    has(name: string): boolean {
        return this.byName.has(name);
    }

    noul(name: string): NoulAnswer {
        return this.expect(name, NoulAnswer, 'noul');
    }

    choice(name: string): ChoiceAnswer {
        return this.expect(name, ChoiceAnswer, 'choice');
    }

    score(name: string): ScoreAnswer {
        return this.expect(name, ScoreAnswer, 'score');
    }

    private expect<T extends Answer>(name: string, expected: new (data: Record<string, unknown>) => T, wanted: string): T {
        const answer = this.answer(name);

        if (!(answer instanceof expected)) {
            throw new TypeSafeError(`Answer "${name}" is a ${answer.type()} answer, not ${wanted}.`);
        }

        return answer;
    }
}

File: python/src/jevlint/typesafe/answers.py
"""
One question's answer.

The SDK hands back typed models; these wrap the payload so a caller can ask what
kind of answer it is, and so an answer type a later API adds is left out instead
of taking the response down with it.
"""
from __future__ import annotations

from typing import Any

from .errors import TypeSafeError


class Answer:
    def __init__(self, data: dict[str, Any]) -> None:
        self._data = data

    def type(self) -> str:
        """The discriminator the API sent."""
        raise NotImplementedError

    def to_dict(self) -> dict[str, Any]:
        return self._data

    @staticmethod
    def from_dict(data: dict[str, Any]) -> Answer | None:
        """The answer class matching the payload's `type`, or None for one not modelled here."""
        return {
            'noul': NoulAnswer,
            'choice': ChoiceAnswer,
            'score': ScoreAnswer,
        }.get(data.get('type'), lambda _: None)(data)

    def _number(self, key: str) -> float:
        value = self._data.get(key)

        if not isinstance(value, (int, float)) or isinstance(value, bool):
            raise TypeSafeError(f'Expected a number for "{key}" in a {self.type()} answer.')

        return float(value)

    def _number_map(self, key: str) -> dict[str, float]:
        values = self._data.get(key) or {}

        if not isinstance(values, dict):
            raise TypeSafeError(f'Expected a map for "{key}" in a {self.type()} answer.')

        found: dict[str, float] = {}

        for name, value in values.items():
            try:
                found[str(name)] = float(value)
            except (TypeError, ValueError):
                found[str(name)] = 0.0

        return found


class NoulAnswer(Answer):
    """A yes/no answer."""

    def type(self) -> str:
        return 'noul'

    def noul(self) -> float:
        """Probability of a yes answer, from zero to one."""
        return self._number('noul')


class ChoiceAnswer(Answer):
    """The label the model selected, with its probability across every label."""

    def type(self) -> str:
        return 'choice'

    def choice(self) -> str:
        choice = self._data.get('choice')

        return choice if isinstance(choice, str) else ''

    def confidence(self) -> float:
        return self._number('confidence')

    def probabilities(self) -> dict[str, float]:
        return self._number_map('probabilities')

    def probability_of(self, label: str) -> float | None:
        """The probability of one label, or None where the model did not report it."""
        return self.probabilities().get(label)


class ScoreAnswer(Answer):
    """An expected score with the rubric it was drawn from."""

    def type(self) -> str:
        return 'score'

    def score(self) -> float:
        """Expected score, which may fall between the integer rubric levels."""
        return self._number('score')

    def nearest_level(self) -> int:
        return round(self.score())

    def confidence(self) -> float:
        return self._number('confidence')

    def legend(self) -> dict[int, Any]:
        legend = self._data.get('legend') or {}

        if not isinstance(legend, dict):
            return {}

        return {int(score): description for score, description in legend.items()}


class Usage:
    """
    Token counts for a request, when the API reports them.

    A provider that reports none is not a call that cost nothing, so the two are
    kept apart and the report says how many calls it could not price.
    """

    def __init__(self, input_tokens: int | None, output_tokens: int | None) -> None:
        self.input_tokens = input_tokens
        self.output_tokens = output_tokens

    @staticmethod
    def from_dict(data: dict[str, Any] | None) -> Usage:
        data = data or {}

        return Usage(_as_number(data.get('input_tokens')), _as_number(data.get('output_tokens')))

    def total_tokens(self) -> int | None:
        """Input and output tokens combined, or None where neither was reported."""
        if self.input_tokens is None and self.output_tokens is None:
            return None

        return (self.input_tokens or 0) + (self.output_tokens or 0)


def _as_number(value: Any) -> int | None:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return None

    return int(value)


class SystemOneResponse:
    """Answers keyed by question name, with the model that produced them and the tokens they cost."""

    def __init__(self, model: str, answers: dict[str, Answer], usage: Usage) -> None:
        self.model = model
        self.usage = usage
        self._answers = answers

    @staticmethod
    def from_dict(data: dict[str, Any]) -> SystemOneResponse:
        answers: dict[str, Answer] = {}
        raw = data.get('answers')

        if isinstance(raw, dict):
            for name, answer in raw.items():
                if not isinstance(answer, dict):
                    continue

                # An answer type a later API adds is left out rather than fatal.
                built = Answer.from_dict(answer)

                if built is not None:
                    answers[str(name)] = built

        model = data.get('model')
        usage = data.get('usage')

        return SystemOneResponse(
            model if isinstance(model, str) else '',
            answers,
            Usage.from_dict(usage if isinstance(usage, dict) else None),
        )

    def answers(self) -> dict[str, Answer]:
        return self._answers

    def answer(self, name: str) -> Answer:
        found = self._answers.get(name)

        if found is None:
            got = ', '.join(self._answers) if self._answers else 'none'

            raise TypeSafeError(f'No answer named "{name}" in the response; got: {got}.')

        return found

    def has(self, name: str) -> bool:
        return name in self._answers

    def noul(self, name: str) -> NoulAnswer:
        return self._expect(name, NoulAnswer, 'noul')

    def choice(self, name: str) -> ChoiceAnswer:
        return self._expect(name, ChoiceAnswer, 'choice')

    def score(self, name: str) -> ScoreAnswer:
        return self._expect(name, ScoreAnswer, 'score')

    def _expect(self, name: str, expected: type, wanted: str) -> Any:
        answer = self.answer(name)

        if not isinstance(answer, expected):
            raise TypeSafeError(f'Answer "{name}" is a {answer.type()} answer, not {wanted}.')

        return answer
